llama_hub/tools/neo4j_db/base.py

```diff
@@ -0,0 +1,146 @@
+from llama_index import Document
+from llama_index.graph_stores import Neo4jGraphStore
+from llama_index.tools.tool_spec.base import BaseToolSpec
+from llama_index.llms.base import LLM, ChatMessage
+
+
+class Neo4jQueryToolSpec(BaseToolSpec):
+    """
+    This class is responsible for querying a Neo4j graph database based on a provided schema definition.
+    """
+
+    def __init__(self, url, user, password, database, llm: LLM):
+        """
+        Initializes the Neo4jSchemaWiseQuery object.
+
+        Args:
+            url (str): The connection string for the Neo4j database.
+            user (str): Username for the Neo4j database.
+            password (str): Password for the Neo4j database.
+            llm (obj): A language model for generating Cypher queries.
+        """
+        try:
+            from neo4j import GraphDatabase
+
+        except ImportError:
+            raise ImportError(
+                "`neo4j` package not found, please run `pip install neo4j`"
+            )
+
+        self.graph_store = Neo4jGraphStore(url=url, username=user, password=password, database=database)
+        self.llm = llm
+
+    def get_system_message(self):
+        """
+        Generates a system message detailing the task and schema.
+
+        Returns:
+            str: The system message.
+        """
+        return f"""
+        Task: Generate Cypher queries to query a Neo4j graph database based on the provided schema definition.
+        Instructions:
+        Use only the provided relationship types and properties.
+        Do not use any other relationship types or properties that are not provided.
+        If you cannot generate a Cypher statement based on the provided schema, explain the reason to the user.
+        Schema:
+        {self.graph_store.schema}
+
+        Note: Do not include any explanations or apologies in your responses.
+        """
+
+    def query_graph_db(self, neo4j_query, params=None):
+        """
+        Queries the Neo4j database.
+
+        Args:
+            neo4j_query (str): The Cypher query to be executed.
+            params (dict, optional): Parameters for the Cypher query. Defaults to None.
+
+        Returns:
+            list: The query results.
+        """
+        if params is None:
+            params = {}
+        with self.graph_store.client.session() as session:
+            result = session.run(neo4j_query, params)
+            output = [r.values() for r in result]
+            output.insert(0, list(result.keys()))
+            return output
+
+    def query_graph_db_as_document(self, neo4j_query, params=None):
```

Comment:
this and the above function are basically the same, one thought is can we just keep one? the original goal of a tool spec is meant more for the LLM agent to use rather than for human use, but it seems like here you were intending this for human use first. 

another thought is we have both, but you only name one in the `spec_functions` array.

Reply:
Yeah, maybe I king missed the whole point of the tool at first and thought perhaps it could be used also for RAG in the main repo of llama_index because it might help get information from the graph in a more accurate way.
anyway, I removed one of the functions.